File: 30. Substring with Concatenation of All Words/SubstringWithConcatenationOfAllWords.cs
public class Solution {
    public IList<int> FindSubstring(string s, string[] words) {
        var result = new List<int>();
        if(string.IsNullOrEmpty(s) || words == null || words.Length == 0)
            return result;

        int wordLength = words[0].Length;
        int wordCount = words.Length;
        int substringLength = wordLength * wordCount;
        
        var wordFrequency = new Dictionary<string, int>();

        foreach(var word in words){
            if(wordFrequency.ContainsKey(word)) wordFrequency[word]++;
            else wordFrequency[word] = 1;
        }
        
        for(int i = 0; i < wordLength; i++) {
            int left = i, count = 0;
            var wordsFound = new Dictionary<string, int>();
            
            for(int j = i; j <= s.Length - wordLength; j += wordLength){
                string word = s.Substring(j, wordLength);
                
                if (wordFrequency.ContainsKey(word)) {
                    if(wordsFound.ContainsKey(word))
                        wordsFound[word]++;
                    else
                        wordsFound[word] = 1;
                    
                    count++;
                    
                    while(wordsFound[word] > wordFrequency[word]) {
                        string leftWord = s.Substring(left, wordLength);
                        wordsFound[leftWord]--;
                        count--;
                        left += wordLength;
                    }
                    
                    if(count == wordCount)
                        result.Add(left);
                } else {
                    wordsFound.Clear();
                    count = 0;
                    left = j + wordLength;
                }
            }
        }

        return result;
    }
}

File: 30. Substring with Concatenation of All Words/substring_with_concatenation_of_all_words.py
class Solution(object):
    def findSubstring(self, s, words):
        from collections import Counter
        
        word_bag = Counter(words) 
        word_len = len(words[0])
        num_words = len(words)
        total_len = word_len * num_words
        res = []
        
        for i in xrange(word_len):
            left = i
            sub_counter = Counter()
            count = 0
            for j in xrange(i, len(s) - word_len + 1, word_len):
                word = s[j:j + word_len]
                if word in word_bag:
                    sub_counter[word] += 1
                    count += 1
                    while sub_counter[word] > word_bag[word]:
                        sub_counter[s[left:left + word_len]] -= 1
                        left += word_len
                        count -= 1
                    if count == num_words:
                        res.append(left)
                else:
                    sub_counter = Counter()
                    left = j + word_len
                    count = 0
                    
        return res